### PR TITLE
Client tls proxy

### PR DESCRIPTION
**Pull request by abhijitherekar**
#### Description
The L7 app proxy is not sending the client certificate for PU-to-PU workflow. This breaks:

1. mtls between PU to PU on L7.
2. Aporeto to Envoy PU http flow.

Fixes: https://github.com/aporeto-inc/aporeto/issues/2370
#### Test plan
*Outline the test plan used to test this change before merging it.*

> Fixes #.

### CODE REVIEW

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "02e7dcb86543de2fcc77f581ebd82d18e61db6a6"
  },
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "61cf6753fa149fc43c54529083046731a4738ace"
  }
]
```

File: controller/internal/enforcer/applicationproxy/http/http.go [hunk 1]
```diff
@@ -155,6 +156,29 @@ func (p *Config) newBaseTLSConfig() *tls.Config {
...
+		SessionTicketsDisabled:   true,
+		GetClientCertificate:     p.GetClientCertificateFunc(),
+		// for now lets make it TLS1.2 as supported max Version.
+		// TODO: Need to test before enabling TLS 1.3, currently TLS 1.3 doesn't work with envoy.
+		MaxVersion: tls.VersionTLS12,
+	}
+}
+
+func (p *Config) GetClientCertificateFunc() func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+	return func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
```
Comment (mheese) [inline controller/internal/enforcer/applicationproxy/http/http.go]: if I see this right, there is no need to return a closure here. Just adjust your function with the right function signature and pass it directly in 166 ... same should be done for `GetCertificateFunc` I guess

Comment (mheese) [reply controller/internal/enforcer/applicationproxy/http/http.go]: also ... as `cert` is  a field that gets updated, you need  a read lock here as well if I'm not mistaken

Comment (abhijitherekar) [reply controller/internal/enforcer/applicationproxy/http/http.go]: for pt. 2 I agree, I will make that change.

For pt. 1 The function is going to get bigger once Envoy changes comes in, so I thought of having it in a different function.

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "7a9fb2235d0fc8335dbbae78815abc84160d3020"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "02e7dcb86543de2fcc77f581ebd82d18e61db6a6"
  }
]
```

File: controller/internal/enforcer/applicationproxy/http/http.go (hunk 1 above)
Comment (mheese) [reply controller/internal/enforcer/applicationproxy/http/http.go]: for pt. 1, this is what I'm saying: change the function signature to:
```
func (p *Config) GetClientCertificateFunc() (*tls.Certificate, error)
```
and assign it like this:
```
GetClientCertificate:     p.GetClientCertificateFunc,
```

Then you can remove the wrapper which builds a closure on every call

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "7a9fb2235d0fc8335dbbae78815abc84160d3020"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "02e7dcb86543de2fcc77f581ebd82d18e61db6a6"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "7a9fb2235d0fc8335dbbae78815abc84160d3020"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "b30028f46e7091af05d168345c4d83ab9f7c9a5c"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "ce252c11fa4d3cd07a4356c6e759c6c90886621c"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "b30028f46e7091af05d168345c4d83ab9f7c9a5c"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "47b2528c63ec3cb668b9f9ab8ce93fa163f938a9"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "b30028f46e7091af05d168345c4d83ab9f7c9a5c"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "3a01ad115473b093bb2f25a80a3de8cbb5a64406"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "b30028f46e7091af05d168345c4d83ab9f7c9a5c"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "b30028f46e7091af05d168345c4d83ab9f7c9a5c"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "9c8c8aeede8195bdba56b620870a99dd8784b0dc"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "f515685cb54cfc71c02756e18fa4e4f7948a5f55"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "71d869652e57fa89c24126aa32b822a0b77886d2"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "785ac8b1eeab7ddb0974e333a6d870b6141ac67a"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "73b25792c00750315d8a51f1d98c1d5f98130ca7"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "29add17bb18a50e7b5615d383fa3d14bf7750d83"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "c5954cb19d64c0849651577307f3b2cd9559f0ff"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "8687a6221a69324c7661f191446e5c45909bfbce"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "c5954cb19d64c0849651577307f3b2cd9559f0ff"
  }
]
```

**Comment by abhijitherekar** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "AppProxy-client-config",
    "component": "trireme-lib",
    "pr-id": "920",
    "commit-sha": "a218f5e0ef549f1c1bd2c197dc5dad609c6e9973"
  },
  {
    "project": "AppProxy-client-config",
    "component": "enforcerd",
    "pr-id": "1502",
    "commit-sha": "c5954cb19d64c0849651577307f3b2cd9559f0ff"
  }
]
```

**Review by mheese — APPROVED** [review]
looks good to me ... exactly as we tested and discussed how it has to be :shipit: 

however, as this is a critical change, I want that @dstiliadis gives his thumbs up on this change as well

**Review by dstiliadis — APPROVED** [review]

File: controller/internal/enforcer/applicationproxy/http/http.go [hunk 2]
```diff
@@ -261,6 +294,7 @@ func (p *Config) RunNetworkServer(ctx context.Context, l net.Listener, encrypted
 		DialContext:         networkDialerWithContext,
 		MaxIdleConnsPerHost: 2000,
 		MaxIdleConns:        2000,
+		ForceAttemptHTTP2:   true,
```
Comment (abhijitherekar) [inline controller/internal/enforcer/applicationproxy/http/http.go:297]: the force is done for the golang issue: https://github.com/golang/go/issues/21336

**Pull request merged**